vision_agent/agent/vision_agent.py

```diff
@@ -154,14 +165,84 @@ def write_plan(
     context = USER_REQ.format(user_request=user_request)
     prompt = PLAN.format(context=context, tool_desc=tool_desc, feedback=working_memory)
     chat[-1]["content"] = prompt
-    return extract_json(model.chat(chat))["plan"]  # type: ignore
+    return extract_json(model.chat(chat))
+
+
+@traceable
+def pick_plan(
+    chat: List[Message],
+    plans: Dict[str, Any],
+    tool_info: str,
+    model: LMM,
+    code_interpreter: CodeInterpreter,
+    verbosity: int = 0,
+) -> Tuple[str, str]:
+    chat = copy.deepcopy(chat)
+    if chat[-1]["role"] != "user":
+        raise ValueError("Last chat message must be from the user.")
+
+    plan_str = format_plans(plans)
+    prompt = TEST_PLANS.format(
+        docstring=tool_info, plans=plan_str, previous_attempts=""
+    )
+
+    code = extract_code(model(prompt))
+    tool_output = code_interpreter.exec_isolation(DefaultImports.prepend_imports(code))
+    tool_output_str = ""
+    if len(tool_output.logs.stdout) > 0:
+        tool_output_str = tool_output.logs.stdout[0]
+
+    if verbosity >= 1:
+        _print_code("Initial code and tests:", code)
+        _LOGGER.info(f"Initial code execution result:\n{tool_output.text()}")
+
+    # retry if the tool output is empty or code fails
+    count = 1
+    while (not tool_output.success or tool_output_str == "") and count < 3:
+        prompt = TEST_PLANS.format(
+            docstring=tool_info,
+            plans=plan_str,
+            previous_attempts=PREVIOUS_FAILED.format(
+                code=code, error=tool_output.text()
+            ),
+        )
+        code = extract_code(model(prompt))
+        tool_output = code_interpreter.exec_isolation(
+            DefaultImports.prepend_imports(code)
+        )
+        tool_output_str = ""
+        if len(tool_output.logs.stdout) > 0:
+            tool_output_str = tool_output.logs.stdout[0]
+
+        if verbosity == 1:
+            _print_code("Code and test after attempted fix:", code)
+            _LOGGER.info(f"Code execution result after attempte {count}")
+
+        count += 1
+
+    user_req = chat[-1]["content"]
+    context = USER_REQ.format(user_request=user_req)
+    # because the tool picker model gets the image as well, we have to be careful with
+    # how much text we send it, so we truncate the tool output to 20,000 characters
+    prompt = PICK_PLAN.format(
+        context=context,
```

Comment:
Is the image part of context ?

Reply:
The image is passed to the model in chat on line 233